FirebaseMLModelDownloader/Sources/ModelInfoRetriever.swift

```diff
@@ -0,0 +1,207 @@
+  @UserDefaultsBacked var path: String?
+
+  /// Initialize model info and create user default keys.
+  init(app: FirebaseApp, name: String, defaults: UserDefaults = .firebaseMLDefaults) {
+    self.name = name
+    self.defaults = defaults
+    let bundleID = Bundle.main.bundleIdentifier ?? ""
+    let defaultsPrefix = "\(bundleID).\(app.name).\(name)"
+    _downloadURL = UserDefaultsBacked(
+      key: "\(defaultsPrefix).model-download-url",
+      storage: defaults
+    )
+    _hash = UserDefaultsBacked(key: "\(defaultsPrefix).model-hash", storage: defaults)
+    _size = UserDefaultsBacked(key: "\(defaultsPrefix).model-size", storage: defaults)
+    _path = UserDefaultsBacked(key: "\(defaultsPrefix).model-path", storage: defaults)
+  }
+}
+
+/// Model info retriever for a model from local user defaults or server.
+class ModelInfoRetriever: NSObject {
+  /// Current Firebase app.
+  var app: FirebaseApp
+  /// Model info associated with model.
+  var modelInfo: ModelInfo?
+  /// Project id.
+  var projectID: String
+  /// Model name.
+  var modelName: String
+  /// Firebase installations.
+  var installations: Installations
+
+  /// Associate model info retriever with current Firebase app, project ID, and model name.
+  init(app: FirebaseApp, projectID: String, modelName: String) {
+    self.app = app
+    self.projectID = projectID
+    self.modelName = modelName
+    installations = Installations.installations(app: app)
+  }
+
+  /// Build custom model object from model info.
+  func buildModel() -> CustomModel? {
+    /// Build custom model only if model info is filled out, and model file is already on device.
+    guard let info = modelInfo, let path = info.path else { return nil }
+    let model = CustomModel(
+      name: info.name,
+      size: info.size,
+      path: path,
+      hash: info.hash
+    )
+    return model
+  }
+}
+
+/// Extension to handle fetching model info from server.
+extension ModelInfoRetriever {
+  /// HTTP request headers.
+  static let fisTokenHTTPHeader: String = "x-goog-firebase-installations-auth"
+  static let hashMatchHTTPHeader: String = "if-none-match"
+  static let bundleIDHTTPHeader: String = "x-ios-bundle-identifier"
+
+  /// HTTP response headers.
+  static let etagHTTPHeader: String = "ETag"
+
+  /// Error descriptions.
+  static let tokenErrorDescription: String = "Error retrieving FIS token."
+  static let selfDeallocatedErrorDescription: String = "Self deallocated."
+  static let missingModelHashErrorDescription: String = "Model hash missing in server response."
+  static let invalidHTTPResponseErrorDescription: String =
+    "Could not get a valid HTTP response from server."
+
+  /// Construct model fetch base URL.
+  var modelInfoFetchURL: URL {
+    var components = URLComponents()
+    components.scheme = "https"
+    components.host = "firebaseml.googleapis.com"
+    components.path = "/v1beta2/projects/\(projectID)/models/\(modelName):download"
+    return components.url!
```

Comment:
No force unwrap should be used in production code.

Reply:
I've added a TODO to handle this in a future PR.